src/wallet/wallet.cpp

```diff
@@ -1775,9 +1839,21 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
             WalletLogPrintf("Still rescanning. At block %d. Progress=%f\n", block_height, progress_current);
         }
 
+        bool skip_block = false;
+        if (fast_rescan_filter) {
+            fast_rescan_filter->UpdateIfNeeded();
+            auto match_result = fast_rescan_filter->MatchesBlock(block_hash);
+            if (match_result.has_value() && !match_result.value()) {
+                // Block is skipped, pretend that we found it successfully
+                result.last_scanned_block = block_hash;
+                result.last_scanned_height = block_height;
+                skip_block = true;
+            }
+        }
```

Comment:
What do you think about putting this in a separate function that returns `bool skip_block`? `CWallet::ScanForWalletTransactions()` is already quite bloated, I think we should try not to add to it?

Reply:
Thought about that already and agree that bloat should be avoided, but couldn't come up with a satisfying function name yet (`CanBlockBeSkippedDueToFastRescanFilter`?) and was concerned that the "fast rescan" functionality in the wallet would become too scattered. One class and once place where to use this class seems to be more compact and self-contained than class + function (that passes the class instance) + another function where this is called.

Any conrete suggestions there? Happy to follow-up. (Another possible variant would be to put the check directly into `FastWalletRescanFilter`, as a method? Not sure.)

Reply:
I like the idea of moving (part of the logic) to `FastWalletRescanFilter`, yeah. What do you think about the below diff? I didn't test it very carefully (tests pass though) but it seems equivalent and imo more readable?

<details>

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index d3f1fb418..fdc5fdc7f 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -282,6 +282,14 @@ public:
         }
     }
 
+    bool BlockNotInFilter(uint256 block_hash)
+    {
+        UpdateIfNeeded();
+        auto block_in_filter{MatchesBlock(block_hash)};
+        return block_in_filter.value_or(true) == false;
+    }
+
+private:
     void UpdateIfNeeded()
     {
         // repopulate filter with new scripts if top-up has happened since last iteration
@@ -300,7 +308,6 @@ public:
         return m_wallet.chain().blockFilterMatchesAny(BlockFilterType::BASIC, block_hash, m_filter_set);
     }
 
-private:
     const CWallet& m_wallet;
     /** Map for keeping track of each active descriptor's last seen end range.
       * This information is used to detect whether new addresses were derived
@@ -1843,16 +1850,7 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
             WalletLogPrintf("Still rescanning. At block %d. Progress=%f\n", block_height, progress_current);
         }
 
-        bool skip_block{false};
-        if (fast_rescan_filter) {
-            fast_rescan_filter->UpdateIfNeeded();
-            auto match_result{fast_rescan_filter->MatchesBlock(block_hash)};
-            if (match_result.has_value() && !match_result.value()) {
-                result.last_scanned_block = block_hash;
-                result.last_scanned_height = block_height;
-                skip_block = true;
-            }
-        }
+        bool skip_block{fast_rescan_filter ? fast_rescan_filter->BlockNotInFilter(block_hash) : false};
 
         // Read block data
         CBlock block;
@@ -1865,7 +1863,10 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
         uint256 next_block_hash;
         chain().findBlock(block_hash, FoundBlock().inActiveChain(block_still_active).nextBlock(FoundBlock().inActiveChain(next_block).hash(next_block_hash)));
 
-        if (!skip_block) {
+        if (skip_block) {
+            result.last_scanned_block = block_hash;
+            result.last_scanned_height = block_height;
+        } else {
             if (!block.IsNull()) {
                 LOCK(cs_wallet);
                 if (!block_still_active) {

```

</details>

Reply:
Would prefer that for a follow-up, could possibly even be a larger one that generally tries to divide up `CWallet::ScanForWalletTransactions()` into more functions to reduce bloat. Just one thing about naming, though I also unfortunately don't have a better idea: `BlockNotInFilter` suggests that this method only retrieves information and as such doesn't have any side-effects (but it does, due to the `UpdateIfNeeded` call).

Reply:
Initially I agreed with your view (and had the same consideration when proposing) about `BlockNotInFilter` having side-effects. However, in this case upon further thinking I'm not sure that's a problem:
- both `BlockNotInFilter()` and `UpdateIfNeeded()` are idempotent, we don't need to be concerned about running it "too often". `UpdateIfNeeded()` only has internal side-effects.
- `UpdateIfNeeded()` is a separate/manual function to improve performance, so we don't continuously need to monitor/update the filter when it's not needed. All it does, is ensure the filter works _as expected_. From that perspective, I don't think `BlockNotInFilter()` has the kind of side-effects we should normally be worried about.

Reply:
Good valid points, I tend to agree that in this case the side-effect is acceptable and we shouldn't worry. With the current version of the PR that introduced logging messages on two cases of the tri-state result (block filter not found or wallet filter matches block), as suggested by Sjors, it get's maybe a bit more difficult to put all this logic into a single method though; the logging would also need to be part of the method, if we want to distinguish between the concrete reason of why we don't skip a block. That said, I'm of course still open though for a potential follow-up that reduces `ScanForWalletTransaction` bloat.